include/ck/tensor_operation/gpu/element/trinary_element_wise_operation.hpp

```diff
@@ -0,0 +1,283 @@
+    {
+        y = type_convert<half_t>(alpha_) * x0 + type_convert<half_t>(beta_) * x1 +
+            type_convert<half_t>(gamma_) * x2;
+    };
+
+    template <>
+    __host__ __device__ constexpr void operator()<half_t, float, half_t, half_t>(
+        half_t& y, const float& x0, const half_t& x1, const half_t& x2) const
+    {
+        y = type_convert<half_t>(alpha_ * x0 + beta_ * ck::type_convert<float>(x1) + gamma_ * x2);
+    };
+
+    float alpha_;
+    float beta_;
+    float gamma_;
+};
+
+struct TriAddRelu
+{
+    template <typename Y, typename X0, typename X1, typename X2>
+    __host__ __device__ constexpr void
+    operator()(Y& y, const X0& x0, const X1& x1, const X2& x2) const;
+
+    template <>
+    __host__ __device__ constexpr void operator()<float, float, float, float>(float& y,
+                                                                              const float& x0,
+                                                                              const float& x1,
+                                                                              const float& x2) const
+    {
+        const float a = x0 + x1 + x2;
+        y             = a > 0.0f ? a : 0.0f;
+    };
+
+    template <>
+    __host__ __device__ constexpr void operator()<double, double, double, double>(
+        double& y, const double& x0, const double& x1, const double& x2) const
+    {
+        const double a = x0 + x1 + x2;
+        y              = a > 0.0 ? a : 0.0;
+    };
+
+    template <>
+    __host__ __device__ constexpr void operator()<half_t, half_t, half_t, half_t>(
+        half_t& y, const half_t& x0, const half_t& x1, const half_t& x2) const
+    {
+        const half_t a = x0 + x1 + x2;
+        y              = a > type_convert<half_t>(0.0f) ? a : type_convert<half_t>(0.0f);
+    };
+
+    template <>
+    __host__ __device__ constexpr void operator()<half_t, float, half_t, half_t>(
+        half_t& y, const float& x0, const half_t& x1, const half_t& x2) const
+    {
+        const float a = x0 + x1 + x2;
+        y             = a > type_convert<half_t>(0.0f) ? a : type_convert<half_t>(0.0f);
+    };
+
+    template <>
+    __host__ __device__ constexpr void operator()<float, float, half_t, float>(
+        float& y, const float& x0, const half_t& x1, const float& x2) const
+    {
+        const float a = x0 + type_convert<float>(x1) + x2;
+        y             = a > 0.0f ? a : 0.0f;
+    };
+};
+
+struct TriAddHardswish
+{
+    template <typename T>
+    __host__ __device__ constexpr void
+    operator()(T& y, const T& x0, const T& x1, const T& x2) const;
+
+    template <>
+    __host__ __device__ constexpr void
+    operator()<float>(float& y, const float& x0, const float& x1, const float& x2) const
+    {
+        float a = x0 + x1 + x2;
+        float b = a + float{3};
+        float c = (b > 0) * (b > 6.0f ? 6.0f : b) * a * 0.166667f;
+        y       = c;
+    };
```

Comment:
I think it's better to make smart type selection instead of duplicating (triplicating? :) ) the code.
Something like this (probably appropriate explicit casts should be added):
```
    template <typename T>
    __host__ __device__ constexpr void
    operator()(T& y, const T& x0, const T& x1, const T& x2) const
    {
        static_assert(!(std::is_same_v<T,double> || std::is_same_v<T,float> || std::is_same_v<T,half_t>));
        using result_type = std::conditional_t<std::is_same_v<T,double>, double, float>;
        result_type  a = x0 + x1 + x2;
        result_type  b = a + result_type{3};
        result_type  c = (b > 0) * (std::min(b, result_type{6})) * a * result_type{0.166667};
        y       = c;
    }
```